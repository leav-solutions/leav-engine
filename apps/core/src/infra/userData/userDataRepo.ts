// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {aql} from 'arangojs';
import {IQueryInfos} from '_types/queryInfos';
import {IUserData} from '_types/userData';
import {IDbService} from '../db/dbService';

const USER_DATA_COLLECTION = 'core_user_data';

export interface IUserDataRepo {
    saveUserData(key: string, value: any, global: boolean, ctx: IQueryInfos): Promise<IUserData>;
    getUserData(keys: string[], global: boolean, ctx: IQueryInfos): Promise<IUserData>;
}

interface IDeps {
    'core.infra.db.dbService'?: IDbService;
}

export default function ({'core.infra.db.dbService': dbService = null}: IDeps = {}): IUserDataRepo {
    return {
        async saveUserData(key: string, value: any, global: boolean, ctx: IQueryInfos): Promise<IUserData> {
            const collection = dbService.db.collection(USER_DATA_COLLECTION);

            const res = await dbService.execute({
                query: aql`
                    UPSERT ${{userId: global ? null : ctx.userId}}
                    INSERT ${{userId: global ? null : ctx.userId, data: {[key]: value}}}
                    UPDATE ${{data: {[key]: typeof value !== 'undefined' ? value : null}}}
                    IN ${collection}
                    OPTIONS { mergeObjects: true, keepNull: false }
                    RETURN NEW`,
                ctx
            });

            return {global, data: {[key]: res[0].data[key]}};
        },
        async getUserData(keys: string[], global: boolean, ctx: IQueryInfos): Promise<IUserData> {
            const collection = dbService.db.collection(USER_DATA_COLLECTION);

            const userData = await dbService.execute({
                query: aql`
                    RETURN KEEP(FIRST(FOR e IN ${collection}
                        FILTER e.userId == ${global ? null : ctx.userId}
                    RETURN e).data, ${keys})`,
                ctx
            });

            return {global, data: userData[0] || {}};
        }
    };
}
