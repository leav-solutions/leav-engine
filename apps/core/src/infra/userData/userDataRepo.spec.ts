// Copyright LEAV Solutions 2017
// This file is released under LGPL V3
// License text available at https://www.gnu.org/licenses/lgpl-3.0.txt
import {Database} from 'arangojs';
import userDataRepo from './userDataRepo';

describe('UserDataRepo', () => {
    const ctx = {
        userId: '1',
        requestId: 'userDataRepoTest'
    };

    test('save user data', async function () {
        const mockDbServ = {
            db: new Database(),
            execute: global.__mockPromise([{data: {test: 'value'}}])
        };

        const udr = userDataRepo({
            'core.infra.db.dbService': mockDbServ
        });

        const res = await udr.saveUserData('test', 'value', false, ctx);

        expect(mockDbServ.execute.mock.calls.length).toBe(1);
        expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery

        expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/UPSERT/);
        expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/INSERT/);
        expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/UPDATE/);
        expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
        expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

        expect(res).toEqual({data: {test: 'value'}, global: false});
    });

    test('get user data', async function () {
        const mockDbServ = {
            db: new Database(),
            execute: global.__mockPromise([{test: 'data'}])
        };

        const udr = userDataRepo({
            'core.infra.db.dbService': mockDbServ
        });

        const res = await udr.getUserData(['test'], false, ctx);

        expect(mockDbServ.execute.mock.calls.length).toBe(1);
        expect(typeof mockDbServ.execute.mock.calls[0][0]).toBe('object'); // AqlQuery

        expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatch(/FILTER/);
        expect(mockDbServ.execute.mock.calls[0][0].query.query).toMatchSnapshot();
        expect(mockDbServ.execute.mock.calls[0][0].query.bindVars).toMatchSnapshot();

        expect(res).toEqual({global: false, data: {test: 'data'}});
    });
});
