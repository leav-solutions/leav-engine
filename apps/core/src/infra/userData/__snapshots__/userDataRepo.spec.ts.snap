// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UserDataRepo get user data 1`] = `
"
                    LET userData = 
                        FIRST(FOR e IN @@value0
                            FILTER e.userId == @value1
                        RETURN e)

                    LET MERGED = MERGE(userData.data || {}, userData.core_data || {})

                    RETURN KEEP(MERGED, @value2)
                "
`;

exports[`UserDataRepo get user data 2`] = `
{
  "@value0": "core_user_data",
  "value1": "1",
  "value2": [
    "test",
  ],
}
`;

exports[`UserDataRepo save user data 1`] = `
"
                    UPSERT @value0
                    INSERT @value1
                    UPDATE @value2
                    IN @@value3
                    OPTIONS { mergeObjects: true, keepNull: false }
                    RETURN NEW"
`;

exports[`UserDataRepo save user data 2`] = `
{
  "@value3": "core_user_data",
  "value0": {
    "userId": "1",
  },
  "value1": {
    "data": {
      "test": "value",
    },
    "userId": "1",
  },
  "value2": {
    "data": {
      "test": "value",
    },
  },
}
`;
